rust/parquet/src/file/reader.rs

```diff
@@ -50,22 +50,52 @@ use crate::util::{io::FileSource, memory::ByteBufferPtr};
 /// Parquet file reader API. With this, user can get metadata information about the
 /// Parquet file, can get reader for each row group, and access record iterator.
 pub trait FileReader {
+    type RowGroupReader: RowGroupReader;
+
     /// Get metadata information about this file.
     fn metadata(&self) -> ParquetMetaDataPtr;
 
     /// Get the total number of row groups for this file.
     fn num_row_groups(&self) -> usize;
 
     /// Get the `i`th row group reader. Note this doesn't do bound check.
-    fn get_row_group(&self, i: usize) -> Result<Box<RowGroupReader>>;
+    fn get_row_group(&self, i: usize) -> Result<Self::RowGroupReader>;
 
     /// Get full iterator of `Row`s from a file (over all row groups).
     ///
+    /// [`Row`](`crate::record::types::Row`) can be used for the type parameter `T` to
+    /// read as untyped rows. A tuple or a struct marked with `#[derive(Record)]` can
+    /// be used to read as typed rows.
+    ///
     /// Iterator will automatically load the next row group to advance.
     ///
     /// Projected schema can be a subset of or equal to the file schema, when it is None,
     /// full file schema is assumed.
-    fn get_row_iter(&self, projection: Option<SchemaType>) -> Result<RowIter>;
+    fn get_row_iter<T>(self, projection: Option<Predicate>) -> Result<RowIter<Self, T>>
+    where
+        T: Record,
+        Self: Sized,
+    {
+        RowIter::from_file(projection, self)
+    }
+}
+
+/// Enable APIs to be generic over values or references.
+impl<'a, R> FileReader for &'a R
```

Comment:
To avoid ownership issues that I encountered after the aformentioned change, `get_row_iter<T>` can now take `&self` (as before), or also now `self`.

`impl<'a, R> FileReader for &'a R where R: FileReader` is what enables it to continue taking the former.